_posts/2016-10-04-bazel-rules-protobuf.md

```diff
@@ -0,0 +1,1007 @@
+simultaneously:
+
+```python
+ proto_compile(
+   name = "pluriproto",
+   protos = [":protos"],
+   langs = [
+       "//cpp",
+       "//csharp",
+       "//closure",
+       "//ruby",
+       "//java",
+       "//java:nano",
+       "//python",
+       "//objc",
+       "//node",
+   ],
+   verbose = 1,
+   with_grpc = True,
+ )
+```
+
+```sh
+$ bazel build :pluriproto
+# ************************************************************
+cd $(bazel info execution_root) && bazel-out/host/bin/external/com_github_google_protobuf/protoc \
+--plugin=protoc-gen-grpc-java=bazel-out/host/genfiles/third_party/protoc_gen_grpc_java/protoc_gen_grpc_java \
+--plugin=protoc-gen-grpc=bazel-out/host/bin/external/com_github_grpc_grpc/grpc_cpp_plugin \
+--plugin=protoc-gen-grpc-nano=bazel-out/host/genfiles/third_party/protoc_gen_grpc_java/protoc_gen_grpc_java \
+--plugin=protoc-gen-grpc-csharp=bazel-out/host/genfiles/external/nuget_grpc_tools/protoc-gen-grpc-csharp \
+--plugin=protoc-gen-go=bazel-out/host/bin/external/com_github_golang_protobuf/protoc_gen_go \
+--descriptor_set_out=bazel-genfiles/examples/proto/pluriproto.descriptor_set \
+--ruby_out=bazel-genfiles \
+--python_out=bazel-genfiles \
+--cpp_out=bazel-genfiles \
+--grpc_out=bazel-genfiles \
+--objc_out=bazel-genfiles \
+--csharp_out=bazel-genfiles/examples/proto \
+--java_out=bazel-genfiles/examples/proto/pluriproto_java.jar \
+--javanano_out=ignore_services=true:bazel-genfiles/examples/proto/pluriproto_nano.jar \
+--js_out=import_style=closure,error_on_name_conflict,binary,library=examples/proto/pluriproto:bazel-genfiles \
+--js_out=import_style=commonjs,error_on_name_conflict,binary:bazel-genfiles \
+--go_out=plugins=grpc,Mexamples/proto/common.proto=github.com/pubref/rules_protobuf/examples/proto/pluriproto:bazel-genfiles \
+--grpc-java_out=bazel-genfiles/examples/proto/pluriproto_java.jar \
+--grpc-nano_out=ignore_services=true:bazel-genfiles/examples/proto/pluriproto_nano.jar \
+--grpc-csharp_out=bazel-genfiles/examples/proto \
+--proto_path=. \
+examples/proto/common.proto
+# ************************************************************
+examples/proto/common_pb.rb
+examples/proto/pluriproto_java.jar
+examples/proto/pluriproto_nano.jar
+examples/proto/common_pb2.py
+examples/proto/common.pb.h
+examples/proto/common.pb.cc
+examples/proto/common.grpc.pb.h
+examples/proto/common.grpc.pb.cc
+examples/proto/Common.pbobjc.h
+examples/proto/Common.pbobjc.m
+examples/proto/pluriproto.js
+examples/proto/Common.cs
+examples/proto/CommonGrpc.cs
+examples/proto/common.pb.go
+examples/proto/common_pb.js
+examples/proto/pluriproto.descriptor_set
+```
+
+The various `*_proto_library` rules (that we'll be using below)
+internally invoke this `proto_compile` rule, then consume the
+generated outputs and compile them with the requisite libraries into
+`.class`, `.so`, `.a` (or whatever) objects.
+
+So let's *make something* already! We'll use bazel and rules_protobuf
+to build a gRPC application.
+
```

Comment:
This is a suggestion to add the content above to that readers can get the source code.